One can use random numbers to model things in the real world. Suppose I wanted to simulate flipping a coin 100 times. The result probably won't be exactly 50 heads but it should be close.

In [75]:
heads=0
for i in range(100):
    x=random.randrange(0,2)    #Creates numbers that are either 0 or 1, a range ends at the end number minus one
    print x,
    if x==1:
        heads = heads +1
print
print "Total number of heads: ",heads
print "Percentage of heads: ", heads/100.0*100.0

1 0 0 1 0 0 1 0 0 1 0 0 1 1 1 0 0 1 1 1 0 0 0 1 0 0 0 0 0 1 0 1 0 1 1 0 0 1 0 0 1 0 0 0 1 0 0 0 1 0 1 0 1 0 0 1 0 0 1 1 1 1 1 1 0 0 1 1 0 1 0 0 0 1 1 1 1 0 0 1 1 1 1 1 1 0 0 0 1 0 0 1 1 0 0 1 1 0 0 1
Total number of heads:  47
Percentage of heads:  47.0


The more trails you perform the more likely the number will be close to 50%.

In [72]:
heads=0
for i in range(10000):
    x=random.randrange(0,2)
    if x==1:
        heads = heads +1
print "Total number of heads: ",heads
print "Percentage of heads: ", heads/10000.0*100.0

Total number of heads:  5014
Percentage of heads:  50.14


## Suppose a city had three levies. The levies are old and each one has a 5% chance of failing in a given year. I'm a city official and I would like to predict the chances of having no failures and therefore no repair costs. The knee jerk guess might be 95%. But actually the chance is (.95)(.95)(.95) which is closer to 86%.

In [10]:
zero_failures=0.95**3.0
print zero_failures

0.857375


The chances of only one failure is (0.95)(0.95)(0.05)+(0.95)(0.05)(0.95)+(0.05)(0.95)(0.95) = 3(0.95)(0.95)(0.05)

In [13]:
one_failures = 0.95*0.95*0.05*3
print one_failures

0.135375


Similiarly the chances of two failures is (0.95)(0.05)(0.05)+(0.05)(0.05)(0.95)+(0.05)(0.95)(0.05) = 3(0.95)(0.05)(0.05)

In [14]:
two_failures=0.95*0.05*0.05*3
print two_failures

0.007125


And the very small probability of three failures is (0.05)(0.05)(0.05)

In [15]:
three_failures=0.05**3
print three_failures

0.000125


And of course the probabilities of all possible outcomes must sum to 1.0

In [17]:
all_events = zero_failures+one_failures+two_failures+three_failures
print all_events

1.0


### But all of this was really exhausting. What if I wrote a short program to simulate this? What if I did a million trials of rolling a 20 sided die three times and calculated the ratios. Will I get similar numbers?

In [56]:
import random
numOfTrials = 1000000
numOfLevies = 3
for k in range(numOfLevies+1):
    events[k] = 0.0            # set the number of events in each catagory to zero to start
                               # need to add one to the number of levies to include 0,1,2,or 3 failures
                               # events[0] is the number of trials with zero failures and so on
for i in range(numOfTrials):
    trial_events=0             #at the beginning of each trial set the number of failures to zero
                               # 0-18 represents no failure, 19 represents failure 
    for j in range(numOfLevies):
        if random.randrange(0,20) > 18:    # roll a 20 sided die (goes from 0 to 19)
            trial_events = trial_events + 1     # each time you roll a 19 that is a failure
    events[trial_events]=events[trial_events]+1    #add result of each trial to total number of trials with that result
print events
events_prob=[]
for k in range(numOfLevies+1):
    events_prob.append(events[k]/numOfTrials)    #divide by the number of trials to convert into probabilities
print events_prob

[857509.0, 135190.0, 7176.0, 125.0]
[0.857509, 0.13519, 0.007176, 0.000125]


In [52]:
#Should be similar to hand calculated probabilities
print zero_failures,one_failures, two_failures, three_failures

0.857375 0.135375 0.007125 0.000125


What if I wanted to change my model? Suppose another engineer states that the probability of failure rate has increased to 10% because of the record breaking freeze thaw cycles the levies underwent last winter. I could redo all the calculations I made in the beginning using 0.1 and 0.9 instead of 0.95 and 0.05. Or I could change a single number in the program and rerun. Instead of the die number being greater than 18, failure is now represented by a number greater than 17.

In [76]:
import random
numOfTrials = 1000000
numOfLevies = 3
for k in range(numOfLevies+1):
    events[k] = 0.0            # set the number of events in each catagory to zero to start
                               # need to add one to the number of levies to include 0,1,2,or 3 failures
                               # events[0] is the number of trials with zero failures and so on
for i in range(numOfTrials):
    trial_events=0             #at the beginning of each trial set the number of failures to zero
                               # 0-18 represents no failure, 19 represents failure 
    for j in range(numOfLevies):
        if random.randrange(0,20) > 17:    # roll a 20 sided die (goes from 0 to 19)
            trial_events = trial_events + 1     # each time you roll an 18 or 19 that is a failure
    events[trial_events]=events[trial_events]+1    #add result of each trial to total number of trials with that result
print events
events_prob=[]
for k in range(numOfLevies+1):
    events_prob.append(events[k]/numOfTrials)    #divide by the number of trials to convert into probabilities
print events_prob

[728631.0, 243398.0, 26962.0, 1009.0]
[0.728631, 0.243398, 0.026962, 0.001009]


Now the probability of no failure has gone down from approximately 86% to 73%

If extra layers of complexity are added, the probabilities can still be calculated by hand but it becomes more and more complicated and laborious to do. There could be more levies, levies that were built at different times and therefore have different probabilities of failure. Different repair costs. Repair vs replace cost decisions. Changing a computer program to take these into account is relatively simple compared to charting out the enormous tree of all possible outcomes.